Update ExtraHop Integration

Dan-at-Extrahop opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Status
Ready

## Related Issues

## Description
New version of the ExtraHop integration (full featured). Complete content package with new powerful commands, real-time incident creation via REST, associated playbooks, enhanced system playbook, end-to-end ticket tracking through a playbook and a field trigger script.  This is an ExtraHop supported integration that has been in the works as a partnership since last year, most recently we've been working with:
- Tyler R
- Matt C
- Prasen S
- Marketing team

## Screenshots
![image](https://user-images.githubusercontent.com/36740533/65576303-8e42bb00-df26-11e9-8e27-7058d2c4c01e.png)

## Related PRs

## Required version of Demisto
4.5

## Does it break backward compatibility?
   - No
       - This is a complete rewrite of the integration which consumes all existing command functionality, but renamed commands, inputs, and outputs. Renamed to be ExtraHop v2.

## Deprecate old ExtraHop integration?
   - Yes

## Must have
- [x] [Tests](https://github.com/Dan-at-Extrahop/content/blob/master/TestPlaybooks/playbook-ExtraHop_v2_Test.yml)
- [x] [Documentation](https://github.com/Dan-at-Extrahop/content/blob/master/Integrations/ExtraHop_v2/ExtraHop_v2_documentation.html)
- [x] Code Review

## Dependencies

## Additional changes

## Technical writer review
Mention and link to the files that require a technical writer review.
- [x] [YAML file](https://github.com/Dan-at-Extrahop/content/blob/master/Integrations/ExtraHop_v2/ExtraHop_v2.yml)
- [x] [CHANGELOG](https://github.com/Dan-at-Extrahop/content/blob/master/Integrations/ExtraHop_v2/CHANGELOG.md)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=4481) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=4481) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=4481) it.</sub>

Dan-at-Extrahop commented:
The ci build is passing with the exception of the intentional change for breaking backwards compatibility, see section in PR description. Thanks

yaakovi commented:
@michalgold please go over the included playbooks.

Dan-at-Extrahop commented on Integrations/ExtraHop_v2/CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+## [Unreleased]
```
This suggests creating the CHANGELOG files: https://github.com/demisto/content/tree/master/docs/release_notes
Where with the auto-generated content come from?

yaakovi replied:
during the release process, new content entity, will have auto-generated CHANGELOG file using their descriptions.

CHANGELOG files should only be updated when the content entity is modified (not created).

Dan-at-Extrahop commented on Integrations/ExtraHop_v2/ExtraHop_v2_documentation.html:
```diff
@@ -0,0 +1,4289 @@
+<p>
```
This was created following these steps: https://github.com/demisto/content/tree/master/docs/integration_documentation
Which creates an HTML document, changing the extension to markdown doesn't make sense?

yaakovi replied:
markdown supports HTML.
when browsing the integration folder in github, the README file is automatically shown under the file list.
this is why we are placing the docs under that file name.

Dan-at-Extrahop commented on Playbooks/playbook-Endpoint_Enrichment_-_Generic_v2.1.yml:
```diff
@@ -792,5 +918,9 @@ outputs:
   description: The device information about the hostname that was enriched using Cylance
     Protect v2.
   type: unknown
+- contextPath: ExtraHop.Device
```
I was just following how CylanceProtect added the contextPath of CylanceProtectDevice on line 917.  Is it reasonable for me to add all 20ish ExtraHop device fields as outputs here?

yaakovi replied:
yes,
i'll fix the `CylanceProtect` after this PR is merged.

Dan-at-Extrahop commented on Integrations/ExtraHop_v2/ExtraHop_v2.py:
```diff
@@ -0,0 +1,943 @@
+except Exception as e:
+    LOG(str(e))
+    LOG.print_log()
+    raise
```
Not sure what to change here, this was written by demisto in the existing version of the ExtraHop integration.  Change raise to return_error or just remove raise?

yaakovi replied:
remove all 3 lines (941-943) and add `return_error` instead.
as you mentioned, we don't like raising errors outside the integration scope.

Dan-at-Extrahop commented on Integrations/ExtraHop_v2/ExtraHop_v2.yml:
```diff
@@ -0,0 +1,3035 @@
+  subtype: python3
+  commands:
+  - name: extrahop-get-alerts
+    arguments: []
```
This is retrieving "alert rules" not alerts, and furthermore this API endpoint does not support limiting, so we have to retrieve all "alert rules" and then the limiting would need to be done with slicing to limit display.  Is this still required? This was pulled directly from the first version of the ExtraHop integration.

Dan-at-Extrahop commented on Integrations/ExtraHop_v2/ExtraHop_v2.yml:
```diff
@@ -0,0 +1,3035 @@
+      defaultValue: "20"
+    - name: offset
+      description: The number of records to skip in the query results.
+    - name: field1
```
This was a design decision tradeoff to be able to auto-populate the values for fields and operators vs. allowing unlimited field searching for power users. It was a tradeoff, but after working with Tyler Rorabaugh he reiterated making it easier for SOC analysts that don't have intimate knowledge of ExtraHop to use.  Curious, using the single argument CSV strategy, how would auto-populating list options work? There are  ~300 fields and ~10 operators, so how would the user know what they are and be able to quickly use them?

Dan-at-Extrahop replied:
Correct

Dan-at-Extrahop commented on Integrations/ExtraHop_v2/ExtraHop_v2.py:
```diff
@@ -0,0 +1,943 @@
+    format_device_with_protocols(device, get_appliance_uuids())
+
+
+def tag_devices_command():
```
The entry is returned in the format_device_with_protocols() function

Dan-at-Extrahop commented on Integrations/ExtraHop_v2/ExtraHop_v2.py:
```diff
@@ -0,0 +1,943 @@
+            start = from_time
+            end = until_time
+        else:
+            return_error("Error when using a fixed time range both from_time and until_time timestamps need to be provided.")
```
This is due to a conditional requirement, if you choose to provide a fixed time range then you need to provide both from and until.  Not sure that I understand why raising an exception here is the right this to do, [the docs state not to raise exceptions.](https://github.com/demisto/content/blob/master/docs/code_conventions/README.MD#return-errors-do-not-raise-them)

yaakovi replied:
![image](https://user-images.githubusercontent.com/30797606/66916956-0b90b580-f025-11e9-93bf-c9e38ba6c329.png)

Dan-at-Extrahop commented on Playbooks/playbook-Endpoint_Enrichment_-_Generic_v2.1.yml:
```diff
@@ -728,8 +729,131 @@ tasks:
+      version: -1
+      name: Get host information from ExtraHop Reveal(x)
+      description: Get host information from ExtraHop Reveal(x).
+      script: 'ExtraHop v2|||extrahop-device-search'
```
Got it, assuming that I should remove this in all of my playbooks added and not just the Demisto curated one, correct?

Dan-at-Extrahop replied:
Got it, assuming that I should remove this in all of my playbooks added and not just the Demisto curated one, correct?

yaakovi replied:
yes

yaakovi commented:
@idovandijk please review the playbooks here

idovandijk commented:
General Points
* Well done, the playbooks look great!
* After configuring your integration, I used the "Test" button which completed successfully. However, when running your test playbook, the very first task resulted in an error:
![image](https://user-images.githubusercontent.com/43602124/67774660-73ec8780-fa66-11e9-98a5-080915d84ccd.png)

so currently the "test" command is misleading. Please fix it to properly test whether the integration works and can be used with the entered credentials.

* Please edit all your different playbooks to use commands that aren't tied to a specific integration. What I mean, is that by doing this: 
![image](https://user-images.githubusercontent.com/43602124/67774684-7ea71c80-fa66-11e9-9e89-0c984f2a1b8f.png)


you are telling it to run `extrahop-device-search` only using "ExtraHop v2" integration. This is unnecessary, and is usually used when one wants to run a general command like `!url` using only a certain integration that supports it (out of many others). Moreover, if you release a new version of ExtraHop, these tasks won't work anymore, because they're tied to ExtraHop v2. It should look like this:
![image](https://user-images.githubusercontent.com/43602124/67774707-8797ee00-fa66-11e9-9025-cc97f9312c8c.png)


* Please change the task names to meet our conventions. You can refer to these playbooks, which meet our standards and are good examples, to see how we name different tasks: `IP Enrichment - Internal - Generic v2`, `Extract Indicators From File - Generic v2`. Note capitalization of letters and the meaningfulness and conciseness of the names. Also, task names shouldn't be the same as the name of the command behind them, they should add a little more insight into what's going to happen, in a few words.
 
* I can't seem to see alerts, been having this loading screen for over a day:
![image](https://user-images.githubusercontent.com/43602124/67774732-92eb1980-fa66-11e9-9f24-aa65e4bd2520.png)


I will be adding additional playbook-specific points in the first line of each playbook file.

Dan-at-Extrahop commented:
@idovandijk Thank you for the review.

1. From the loading screen you posted, I noticed that your ExtraHop Reveal(x) instance needs to be upgraded to 7.8 which is the minimum version for this integration. Here's a guide to upgrade, follow the [Upgrade the firmware](https://docs.extrahop.com/7.8/firmware-upgrade/) steps.
2. It looks like you aren't using the latest commit from this branch, I implemented @yaakovi suggestion on removing "ExtraHop v2" version identifiers from all playbook tasks, and I've also updated error messages so that any task errors will be more verbose.  From your Test playbook error screenshot I can see the old version of the errors, please pull the latest commit here and try again.

I'm unable to see the playbook-specific points that you added, please advise where to see your comments. Yes, I'll email you now to setup a time for a call, thank you.

Dan-at-Extrahop commented:
@idovandijk all playbook task names have been updated to meet your conventions.  I also improved the test playbook.

idovandijk commented:
> @idovandijk Thank you for the review.
> 
> 1. From the loading screen you posted, I noticed that your ExtraHop Reveal(x) instance needs to be upgraded to 7.8 which is the minimum version for this integration. Here's a guide to upgrade, follow the [Upgrade the firmware](https://docs.extrahop.com/7.8/firmware-upgrade/) steps.
> 2. It looks like you aren't using the latest commit from this branch, I implemented @yaakovi suggestion on removing "ExtraHop v2" version identifiers from all playbook tasks, and I've also updated error messages so that any task errors will be more verbose.  From your Test playbook error screenshot I can see the old version of the errors, please pull the latest commit here and try again.
> 
> I'm unable to see the playbook-specific points that you added, please advise where to see your comments. Yes, I'll email you now to setup a time for a call, thank you.

I'm sorry, the comments didn't go through, it should appear now. I saw that you pushed some code fixes but I think most of the playbook comments should stay valid. If any comment is outdated please disregard it and continue to the next one. As for the call, I'll get back to you ASAP.

yaakovi commented:
@kirbles19 
please go over the:
integration yml: https://github.com/Dan-at-Extrahop/content/edit/master/Integrations/ExtraHop_v2/ExtraHop_v2.yml?pr=%2Fdemisto%2Fcontent%2Fpull%2F4481

integration description:
https://github.com/Dan-at-Extrahop/content/edit/master/Integrations/ExtraHop_v2/ExtraHop_v2_description.md?pr=%2Fdemisto%2Fcontent%2Fpull%2F4481

generic playbook change log:
https://github.com/Dan-at-Extrahop/content/edit/master/Playbooks/playbook-Endpoint_Enrichment_-_Generic_v2.1_CHANGELOG.md?pr=%2Fdemisto%2Fcontent%2Fpull%2F4481

script yml:
https://github.com/Dan-at-Extrahop/content/edit/master/Scripts/ExtraHopTrackIncidents/ExtraHopTrackIncidents.yml?pr=%2Fdemisto%2Fcontent%2Fpull%2F4481

richardbluestone commented:
@yaakovi I'm going through the integration yml file and I can't review it properly as I don't understand  the context output descriptions. For example, "ExtraHop.Record.Source.to" is defined as "to".  ExtraHop.Record.Source.till is defined as "till".  There are 100's of these. I need more information to understand.

Dan-at-Extrahop commented:
@richardbluestone Thanks for the review, I'm happy to answer any questions to assist. These ExtraHop.Record.Source outputs refer specifically to only our extrahop-query-records command, which is a minor part of the integration used to pull transactional records out of ExtraHop. The descriptions present are the current level of detail available in our product today, since these fields are used generally to cover all 60+ record types. We do not believe this will be an issue for our joint customers, since as mentioned this is the level of detail available for records in ExtraHop and it's understandable in the context of a particular record. Beyond these context outputs, all of the commands and arguments have full descriptions, and the rest of the context outputs have detailed descriptions available. I appreciate your review to get this approved in time for the upcoming release.

Dan-at-Extrahop commented:
@idovandijk all playbook changes have been pushed. @yaakovi the two outstanding build issues are:

- It does not like the mobile layout. Do I need to remove it?
- It does not like keyword ticket in our incident field named detectionticketed.  It seems like this should only complain if we tried to make an incident field that is explicitly named ticket, but that's not the case here.

Lastly just a reminder note for whatever needs to be done to also deprecate the existing ExtraHop integration when this new ExtraHop Reveal(x) integration is released.

This should be all of the changes required for us to merge in time for 11/15 release, with the exception of the review from @richardbluestone.

idovandijk commented:
@yaron-libman Need your review for descriptions of playbooks and tasks, and task names :)
